runtime/runtime-params-estimator/src/estimator_context.rs

```diff
@@ -149,3 +149,9 @@ impl Testbed<'_> {
         }
     }
 }
+
+impl<'a> Drop for Testbed<'a> {
+    fn drop(&mut self) {
+        self.inner.stop_prefetching_threads();
```

Comment:
Let's propagate this drop to `ShardTriesInner`? Joining thread on drop should be responsibility of the entity which spawns the threads. As a rule-of-thumb, every spawn thread should have its' join. 

Perhaps also use something like

```
struct ShardTriesInner {
    /// Prefetcher state, such as IO threads, per shard.
    prefetchers: RwLock<HashMap<ShardUId, (PrefetchApi, Vec<JoinHandle<()>>)>>,
}
```

?

The funky `Clone`  impl is funky, better if we can separate `Prefetcher` and `PrefetcherHandle` at the type level.

Reply:
Yeah, 100% agree that this clone thing is too funky, thanks for pointing it out.

However, `ShardTriesInner::drop` seems like the wrong place to me. At least I failed in my attempt to make it work.

The problem is that we really must close the crossbeam channel first and only then join the threads. Joining when dropping  `Testbed` only works because the testbed itself outlives all places that could hold a clone of a channel sender, which is not true for `ShardTriesInner`.

The sender is stored inside `PrefetchApi`, which in turn is created by `ShardTriesInner`. `PrefetchApi` is cloned around into every instance of `Trie` and that lives in various other structs, such as `TrieUpdate`. And unlike `Testbed`, `ShardTriesInner` is not guaranteed to outlive all of those. Thus, joining when `ShardTriesInner` is dropped results in deadlocks as the background threads can still be waiting on an open channel.

I have now attempted to apply your suggestion in a slightly different way. Definitive ownership of the channel sender AND the join handles is now given to the clonable struct `WorkQueue`, which itself is stored inside `PrefetchApi`. This way all clones of the sender also clone the `Arc<Vec<JoinHandle>>`. When the last instance that combination is dropped, it is safe to join the threads.

I am still not 100% happy with my implementation, though. It is still gimmicky and there are too many nested structs for my taste. But it's the truest representation of ownership that I could come up with right now.^^ Ideas for improvements are welcome. :)